File: PGBooks/PGBooks/Sources/UI/Common/Detail/BookDetailController+Transition.m
//
//  BookDetailController+Transition.m
//  PGBooks
//
//  Created by ipagong on 15/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

#import "BookDetailController+Transition.h"
#import "BookSimpleCell.h"
#import "PGBook.h"
#import "UIViewController+Utils.h"

#define TAG_TRANSITION_CONTAINER    1000
#define TAG_TRANSITION_COVER        1001
#define TAG_TRANSITION_BG           1002
#define TAG_TRANSITION_CELL         1003

@implementation BookDetailController (Transition) 

#pragma mark - Destinationable

- (void)appearWhenPresentWithStatus:(enum PGStatus)status
                           animator:(PGAnimator * _Nonnull)animator
                            context:(PGContext * _Nonnull)context {
    switch (status) {
        case PGStatusPrepare:
        {
            [context.container addSubview:[context.opposite.view snapshotViewAfterScreenUpdates:YES]];
            
            BookSimpleCell * cell = [self originalCellWithContext:context];
            
            self.coverView.alpha = 0.0;
            
            UIView *container = [self.view snapshotViewAfterScreenUpdates:YES];
            container.frame = [cell convertRect:cell.bounds toView:context.container];
            container.tag   = TAG_TRANSITION_CONTAINER;
            container.alpha = 0.0;
            [context.container addSubview:container];
            
            UIView *cover = [cell.coverView snapshotViewAfterScreenUpdates:YES];
            cover.frame = [cell.coverView convertRect:cell.coverView.bounds toView:context.container];
            cover.tag   = TAG_TRANSITION_COVER;
            [context.container addSubview:cover];
            
            break;
        }
        case PGStatusDoing:
        {
            UIView *container = [context.container viewWithTag:TAG_TRANSITION_CONTAINER];
            container.frame = context.target.view.frame;
            container.alpha = 1.0;
            
            UIView *cover = [context.container viewWithTag:TAG_TRANSITION_COVER];
            cover.frame = [self.coverView convertRect:self.coverView.bounds toView:context.container];
            
            break;
        }
        case PGStatusDone:
        case PGStatusCancel:
        {
            [context.container addSubview:context.target.view];
            self.coverView.alpha = 1.0;
            [[context.container viewWithTag:TAG_TRANSITION_CONTAINER] removeFromSuperview];
            [[context.container viewWithTag:TAG_TRANSITION_COVER] removeFromSuperview];
            
            break;
        }
    }
}

- (void)disppearWhenDismissWithStatus:(enum PGStatus)status
                             animator:(PGAnimator * _Nonnull)animator
                              context:(PGContext * _Nonnull)context {
    switch (status) {
        case PGStatusPrepare:
        {
            [context.container addSubview:context.opposite.view];
            
            BookSimpleCell * cell = [self originalCellWithContext:context];
            
            if (cell) { self.coverView.alpha = 0.0; }
            
            UIView *bg = [UIView new];
            bg.frame = context.container.frame;
            bg.backgroundColor = [UIColor whiteColor];
            bg.tag = TAG_TRANSITION_BG;
            [context.container addSubview:bg];
            
            UIView *container = [self.view snapshotViewAfterScreenUpdates:YES];
            container.frame = context.container.frame;
            container.tag = TAG_TRANSITION_CONTAINER;
            [context.container addSubview:container];
            
            cell.coverView.alpha = 0.0;
            UIView *sCell = [cell snapshotViewAfterScreenUpdates:YES];
            sCell.frame = [cell convertRect:cell.bounds toView:context.container];
            sCell.alpha = 0.0;
            sCell.tag = TAG_TRANSITION_CELL;
            [context.container addSubview:sCell];
            
            cell.coverView.alpha = 1.0;
            UIView *cover = [cell.coverView snapshotViewAfterScreenUpdates:YES];
            cover.frame = [self.coverView convertRect:self.coverView.bounds toView:context.container];
            cover.tag = TAG_TRANSITION_COVER;
            [context.container addSubview:cover];
            
            break;
        }
        case PGStatusDoing:
        {
            BookSimpleCell * cell = [self originalCellWithContext:context];
            
            UIView *bg        = [context.container viewWithTag:TAG_TRANSITION_BG];
            UIView *container = [context.container viewWithTag:TAG_TRANSITION_CONTAINER];
            UIView *sCell     = [context.container viewWithTag:TAG_TRANSITION_CELL];
            UIView *cover     = [context.container viewWithTag:TAG_TRANSITION_COVER];
            
            if (cell) {
                bg.frame        = [cell convertRect:cell.bounds toView:context.container];
                container.frame = [cell convertRect:cell.bounds toView:context.container];
                sCell.frame     = [cell convertRect:cell.bounds toView:context.container];
                cover.frame     = [cell.coverView convertRect:cell.coverView.bounds toView:context.container];
            } else {
                CGRect rect = CGRectMake(0,
                                         context.container.frame.size.height,
                                         context.container.frame.size.width,
                                         context.container.frame.size.height/2);
                bg.frame        = rect;
                container.frame = rect;
                sCell.frame     = rect;
                cover.frame     = rect;
            }
            
            [UIView addKeyframeWithRelativeStartTime:0.8 relativeDuration:1.0 animations:^{ bg.alpha = 0.0; }];
            [UIView addKeyframeWithRelativeStartTime:0.0 relativeDuration:0.5 animations:^{ container.alpha = 0.0; }];
            [UIView addKeyframeWithRelativeStartTime:0.5 relativeDuration:1.0 animations:^{ sCell.alpha = 1.0; }];
            
            break;
        }
        case PGStatusDone:
        case PGStatusCancel:
        {
            for (UIView *view in self.stackContainer.subviews) { view.alpha = 1.0; }
            self.coverView.alpha = 1.0;
            [[context.container viewWithTag:TAG_TRANSITION_BG] removeFromSuperview];
            [[context.container viewWithTag:TAG_TRANSITION_CONTAINER] removeFromSuperview];
            [[context.container viewWithTag:TAG_TRANSITION_CELL] removeFromSuperview];
            [[context.container viewWithTag:TAG_TRANSITION_COVER] removeFromSuperview];
            
            break;
        }
    }
}

#pragma mark - utils

- (nullable BookSimpleCell *)originalCellWithContext:(PGContext * _Nonnull)context {
    UIViewController<BookDetailTransitionTargetProtocol> *vc = (UIViewController<BookDetailTransitionTargetProtocol> *)context.opposite.lastChildren;
    
    if (vc == nil) { return nil; }
    if ([vc conformsToProtocol:@protocol(BookDetailTransitionTargetProtocol)] == NO) { return nil; }
    if ([vc respondsToSelector:@selector(tableview)] == NO) { return nil; }
    
    for (BookSimpleCell *visible in [vc.tableview visibleCells]) {
        if ([visible isKindOfClass:[BookSimpleCell class]] == NO)  { continue; }
        if ([visible.key isEqualToString:self.book.isbn13] == YES) { return visible; }
    }
    
    return nil;
}

@end

File: PGBooks-swift/PGBooks-swift/Sources/UI/Common/Detail/BookDetailController+Transition.swift
//
//  BookDetailController+Transition.swift
//  PGBooks-swift
//
//  Created by ipagong on 18/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

import UIKit
import PGTransitionKit

extension BookDetailController : AnimatorDestinationable {
    func appearWhenPresent(status: Status, animator: Animator, context: Context) {
        switch status {
        case .prepare:
            guard let fromSnap = context.opposite.view.snapshotView(afterScreenUpdates: true) else { return }
            
            context.container.addSubview(fromSnap)
            
            guard let cell = self.originalCell(self.book.isbn13, context: context) else { return }
            
            self.coverView.alpha = 0.0
            
            if let container = self.view.snapshotView(afterScreenUpdates: true) {
                container.frame = cell.convert(cell.bounds, to: context.container)
                container.tag   = Constant.Tag.container.rawValue
                container.alpha = 0.0
                context.container.addSubview(container)
            }
            
            if let cover = cell.coverView.snapshotView(afterScreenUpdates: true) {
                cover.frame = cell.coverView.convert(cell.coverView.bounds, to: context.container)
                cover.tag   = Constant.Tag.cover.rawValue
                context.container.addSubview(cover)
            }
            
        case .doing:
            let container = context.container.viewWithTag(Constant.Tag.container.rawValue)
            container?.frame = context.target.view.frame
            container?.alpha = 1.0
            
            let cover = context.container.viewWithTag(Constant.Tag.cover.rawValue)
            cover?.frame = self.coverView.convert(self.coverView.bounds, to: context.container)
            
        case .done, .cancel:
            context.container.addSubview(context.target.view)
            self.coverView.alpha = 1.0
            context.container.viewWithTag(Constant.Tag.container.rawValue)?.removeFromSuperview()
            context.container.viewWithTag(Constant.Tag.cover.rawValue)?.removeFromSuperview()
        }
    }
    
    func disppearWhenDismiss(status: Status, animator: Animator, context: Context) {
        switch status {
        case .prepare:
            context.container.addSubview(context.opposite.view)
            
            let cell = self.originalCell(self.book.isbn13, context: context)
            
            if let _ = cell { self.coverView.alpha = 0.0 }
            
            let bg = UIView()
            bg.frame = context.container.frame
            bg.backgroundColor = UIColor.white
            bg.tag = Constant.Tag.bg.rawValue
            context.container.addSubview(bg)
            
            if let container = self.view.snapshotView(afterScreenUpdates: true) {
                container.frame = context.container.frame
                container.tag = Constant.Tag.container.rawValue
                context.container.addSubview(container)
            }
            
            cell?.coverView.alpha = 0.0
            if let sCell = cell?.snapshotView(afterScreenUpdates: true) {
                sCell.frame = cell?.convert(cell?.bounds ?? .zero, to: context.container) ?? .zero
                sCell.alpha = 0.0
                sCell.tag = Constant.Tag.cell.rawValue
                context.container.addSubview(sCell)
            }
            
            cell?.coverView.alpha = 1.0
            if let cover = cell?.coverView.snapshotView(afterScreenUpdates: true) {
                cover.frame = self.coverView.convert(self.coverView.bounds, to: context.container)
                cover.tag = Constant.Tag.cover.rawValue
                context.container.addSubview(cover)
            }
            
        case .doing:
            let cell = self.originalCell(self.book.isbn13, context: context)
            
            let bg        = context.container.viewWithTag(Constant.Tag.bg.rawValue )
            let container = context.container.viewWithTag(Constant.Tag.container.rawValue )
            let sCell     = context.container.viewWithTag(Constant.Tag.cell.rawValue )
            let cover     = context.container.viewWithTag(Constant.Tag.cover.rawValue )
            
            if let cell = cell {
                bg?.frame        = cell.convert(cell.bounds, to: context.container)
                container?.frame = cell.convert(cell.bounds, to: context.container)
                sCell?.frame     = cell.convert(cell.bounds, to: context.container)
                cover?.frame     = cell.coverView.convert(cell.coverView.bounds, to: context.container)
            } else {
                let rect = CGRect.init(x: 0,
                                       y: context.container.frame.height,
                                       width: context.container.frame.width,
                                       height: context.container.frame.height/2)
                
                bg?.frame        = rect
                container?.frame = rect
                sCell?.frame     = rect
                cover?.frame     = rect
            }
            
            UIView.addKeyframe(withRelativeStartTime: 0.8, relativeDuration: 1.0) { bg?.alpha = 0.0 }
            UIView.addKeyframe(withRelativeStartTime: 0.0, relativeDuration: 0.5) { container?.alpha = 0.0 }
            UIView.addKeyframe(withRelativeStartTime: 0.5, relativeDuration: 1.0) { sCell?.alpha = 1.0 }
            
        case .done, .cancel:
            self.stackContainer.subviews.forEach { $0.alpha = 1.0 }
            self.coverView.alpha = 1.0
            context.container.viewWithTag(Constant.Tag.bg.rawValue)?.removeFromSuperview()
            context.container.viewWithTag(Constant.Tag.container.rawValue)?.removeFromSuperview()
            context.container.viewWithTag(Constant.Tag.cell.rawValue)?.removeFromSuperview()
            context.container.viewWithTag(Constant.Tag.cover.rawValue)?.removeFromSuperview()
            
        }
    }
    
    fileprivate func originalCell(_ key:String, context:Context) -> BookSimpleCell? {
        guard let vc = context.opposite.lastChildren as? BookDetailTransitionable else { return nil }
        return vc.tableview.visibleCells.compactMap{ $0 as? BookSimpleCell }.filter{ $0.data?.isbn13 == key }.first
    }
    
    struct Constant {
        enum Tag : Int {
            case container = 1000
            case cover     = 1001
            case bg        = 1002
            case cell      = 1003
        }
    }
    
}
